### ISSUE TITLE
torch.neuron.DataParallel returns incomplete result for model with dict input when split_size is less than default

### ISSUE THREAD

**Issue by BayMinimum** [aws-neuron/aws-neuron-sdk#620]
Hi, I’m using torch.neuron.DataParallel to utilize all neuroncores in an inf1 instance. It works fine when when DataParallel splits the input into (# of neuroncores) splits, but when it is instructed to split the input into more splits of smaller size by setting `split_size` to less than `batch_size // num_cores` to take full advantage of multiple workers per neuroncore (`num_workers = num_threads * 2`), the model silently returns incomplete result with batch dimension less than input batch dimension. While digging into the issue, I've found that dictionary input is not split properly.

## Code to reproduce

* Package version
  * `torch==1.8.1`
  * `torch-neuron==1.8.1.2.5.8.0`

```python
import torch
import torch.neuron
from torch import nn

class Model(nn.Module):
    def forward(self, d):
        # d: dictionary with schema {"foo": Tensor[batch_size, 1], "bar": Tensor[batch_size, 1]}
        return d["foo"] + d["bar"]

model = Model()
example_input = {"foo": torch.rand(1, 1), "bar": torch.rand(1, 1)}
neuron_model = torch.neuron.trace(model, example_inputs=[example_input], fallback=True, dynamic_batch_size=False)
neuron_dp_model = torch.neuron.DataParallel(model, device_ids=[0, 1, 2, 3])  # 4 neuroncores
neuron_dp_model.disable_dynamic_batching()
neuron_dp_model.num_workers = 8
neuron_dp_model.split_size = 1  # For batch_size=8, default split_size is 8 // 4 = 2

batch_input = {"foo": torch.rand(8, 1), "bar": torch.rand(8, 1)}
batch_output = neuron_dp_model(batch_input)
print(batch_output.shape)  # Output: torch.Size([4, 1])
assert batch_output.shape[0] == 8  # assertion fails; actual: 4
```

## Suspected problem in data_parallel.py

```python
def scatter(inputs, device_ids, split_size_, dim=0):
    r"""
    Breaks the `inputs` into a tuple of chunks of the input on `dim`.
    Duplicates references to objects in `inputs` that are not tensors.
    """
    def scatter_map(obj):
        if isinstance(obj, torch.Tensor):
            num_devices = len(device_ids)
            num_batches = obj.shape[dim]
            split_size,remainder = _get_split_size(split_size_, num_devices, dim, obj)

            ### omitted for readability ###

        if _is_namedtuple(obj):
            return [type(obj)(*args) for args in zip(*map(scatter_map, obj))]
        if isinstance(obj, tuple) and len(obj) > 0:
            return list(zip(*map(scatter_map, obj)))
        if isinstance(obj, list) and len(obj) > 0:
            return [list(i) for i in zip(*map(scatter_map, obj))]
        if isinstance(obj, dict) and len(obj) > 0:
            return [type(obj)(i) for i in zip(*map(scatter_map, obj.items()))]  # HERE(1)
        return [obj for nc in device_ids]  # and HERE(2)

    try:
        res = scatter_map(inputs)
    finally:
        scatter_map = None
    return res
```

In HERE(1), `map(scatter_map, obj.items())` is called with `obj = {"foo": torch.rand(8, 1), "bar": torch.rand(8, 1)}`

→ `scatter_map(("foo", torch.rand(8, 1)))` called

→ `scatter_map("foo")` called, `["foo", "foo", "foo", "foo"]` (**4x** repeat for 4 neuroncores) is returned in HERE(2)

`scatter_map(torch.rand(8, 1))` called, `[Tensor[1, 1], ..., Tensor[1, 1]]` (split into **8** tensors) is returned

→ `[("foo", Tensor[1, 1]), ("foo", Tensor[1, 1]), ("foo", Tensor[1, 1]), ("foo", Tensor[1, 1])]` (length **4 instead of 8 intended**) is returned, since python’s build-in `zip` does not raise error on inputs with different lengths; instead, it breaks on shortest input

## Suggested fixes

- Change HERE(1) to `return [type(obj)(zip(obj.keys(), i)) for i in zip(*map(scatter_map, obj.values()))]` . In this way, `scatter_map` will be recursively applied only on `obj`'s values, instead of its keys.
    - This change is sufficient to fix above snippet for reproduction.
- Change HERE(2) to duplicate other objects by number of splits, instead of number of neuroncores.

**Comment by jluntamazon** [comment]
Hi @BayMinimum:

Thank you for reporting this issue and providing detailed steps to reproduce the problem! We will provide an update on this ticket once we have implemented a fix. The fix should come out in an upcoming release.

**Comment by jeffhataws** [comment]
@BayMinimum, we have implemented a fix and it will be available in an upcoming Neuron SDK release for inf1.

**Comment by BayMinimum** [comment]
Thank you! I'm looking forward to the upcoming release.
